examples/cpp/PineAPPL.hpp

```diff
@@ -0,0 +1,257 @@
+    void set_bool(const std::string& key, const bool value) const { pineappl_keyval_set_bool(this->raw, key.c_str(), value); }
+    void set_int(const std::string& key, const int value) const { pineappl_keyval_set_int(this->raw, key.c_str(), value); }
+    void set_string(const std::string& key, const std::string& value) const { pineappl_keyval_set_string(this->raw, key.c_str(), value.c_str()); }
+    ///@}
+
+    /** @name getter */
+    ///@{
+    double get_double(const std::string& key) const { return pineappl_keyval_double(this->raw, key.c_str()); }
+    bool get_bool(const std::string& key) const { return pineappl_keyval_bool(this->raw, key.c_str()); }
+    int get_int(const std::string& key) const { return pineappl_keyval_int(this->raw, key.c_str()); }
+    std::string get_string(const std::string& key) const { return pineappl_keyval_string(this->raw, key.c_str()); }
+    ///@}
+
+};
+
+/** @brief Entry in luminosity function  */
+struct LumiEntry {
+    /** @brief first parton id */
+    int32_t pid1;
+
+    /** @brief second parton id */
+    int32_t pid2;
+
+    /** @brief relative weight */
+    double weight;
+};
+
+/** @brief Luminosity function */
+struct Lumi {
+    /** @brief underlying raw object */
+    pineappl_lumi *raw;
+
+    /** @brief constructor */
+    Lumi(){ this->raw = pineappl_lumi_new(); }
+
+    /** @brief destructor */
+    ~Lumi(){ pineappl_lumi_delete(this->raw); }
+
+    /** @brief number of elements */
+    size_t count() const { return pineappl_lumi_count(this->raw); }
+
+    /**
+     * @brief add a luminosity function
+     * @param c luminosity function
+     */
+    void add(const std::vector<LumiEntry>& c) const {
+        size_t n = 0;
+        int32_t *pids = new int32_t;
+        double *weights = new double;
+        for (auto it = c.cbegin(); it != c.cend(); ++it, ++n) {
+            pids[2*n] = it->pid1;
+            pids[2*n+1] = it->pid2;
+            weights[n] = it->weight;
+        }
+        pineappl_lumi_add(this->raw, n, pids, weights);
+        delete pids;
+        delete weights;
+    }
+
+    /**
+     * @brief Returns the number of combinations of the luminosity function `lumi` for the specified entry.
+     * @param entry position in lumi
+     */
+    size_t combinations(size_t entry) const { return pineappl_lumi_combinations(this->raw, entry); }
+};
+
+/** @brief Coupling powers for each grid. */
+struct Order {
+    /** @brief Exponent of the strong coupling. */
+    uint32_t alphas;
+
+    /** @brief Exponent of the electromagnetic coupling. */
+    uint32_t alpha;
+
+    /** @brief Exponent of the logarithm of the scale factor of the renomalization scale. */
+    uint32_t logxir;
+
+    /** @brief Exponent of the logarithm of the scale factor of the factorization scale. */
+    uint32_t logxif;
+};
+
+class Grid {
+    /** @brief number of orders */
+    size_t n_orders = 0;
```

Comment:
Remove `n_orders`, `n_bins` and `n_lumis` as member variables and instead make them local to `Grid::convolute_with_one`. The problem is that there are operations that could possibly modify them iternally in `raw` and then they are out-of-sync in C++.

Reply:
fine - then this becomes also a struct: https://github.com/N3PDF/pineappl/pull/178/files#r974460355